numpy/ma/core.py

```diff
@@ -3882,22 +3882,73 @@ def __repr__(self):
             name = self._baseclass.__name__
 
         is_long = self.ndim > 1
-        is_structured = bool(self.dtype.names)
-
-        parameters = dict(
-            name=name,
-            nlen=" " * len(name),
-            data=str(self),
-            mask=str(self._mask),
-            fill=str(self.fill_value),
-            dtype=str(self.dtype)
+
+        # 2016-11-19: Demoted to legacy format
+        if np.get_printoptions()['legacy'] == '1.13':
+            parameters = dict(
+                name=name,
+                nlen=" " * len(name),
+                data=str(self),
+                mask=str(self._mask),
+                fill=str(self.fill_value),
+                dtype=str(self.dtype)
+            )
+            is_structured = bool(self.dtype.names)
+            key = '{}_{}'.format(
+                'long' if is_long else 'short',
+                'flx' if is_structured else 'std'
+            )
+            return _legacy_print_templates[key] % parameters
+
+        prefix = 'masked_{}('.format(name)
+
+        dtype_needed = (
+            not np.core.arrayprint.dtype_is_implied(self.dtype) or
+            np.all(self.mask) or
+            self.size == 0
```

Comment:
Plain ndarrays also have a special case for size-0, more than 1 dimension:
```python
>>> np.empty((0,1))
array([], shape=(0, 1), dtype=float64)
```
But I'm a bit ambivalent about that behavior since shape isn't a valid `array` keyword. But I don't have a better way of signalling that the shape is >1d.

Compare to `np.ma.empty((0,1))`.

Reply:
I just checked, that behavior goes back to [before numpy existed](https://github.com/numpy/numpy/commit/618a8d472a36506eb5848009f10585f7c63da799#diff-da963b1dd9ccb384fce15478f5b49764R150)!

Reply:
An idea: Maybe print `np.empty((0, 1))` as `np.empty((0, 1), dtype=float64)`?

Reply:
That's pretty neat. I'm on board.

I'll try it out as an extra commit in the newline PR.

Reply:
Please don't - I want to get the newline PR in as is so that I can rebase this

Reply:
Or actually, we can try in a new PR to avoid cluttering things.

Reply:
Oh ok I didn't see your rely. Agreed.

Reply:
You can always create a new PR with common commits with the old PR - if you do that, and the old PR is merged, then you can avoid having to rebase by switching the "base branch" twice in the github UI.

Reply:
I suspect that some rebasing will be needed anyway since you'll touch the typename code... we'll see.